### ISSUE TITLE
Proxy doesn't work in some situation

### ISSUE THREAD

**Issue by intari** [httptoolkit/httptoolkit#248]
Situation:
HTTP toolkit has it's proxy configured as http://192.168.88.83:7777 (
.83 is VM which runs CheckPoint VPN and proxy server (3proxy).

When android app wants to connect to existing.domain.com (".com" domain exist in global DNS, "existing.domain.com" is only accessible via Checkpoint but can be resolved) - proxy is being used, trafic is being set to 192.168.88.83 (it's VM which routes traffic via VPN to correct place).
When android app wants to connect to a.b.c.d.existing.domain.zsz (".zsz" domain does not exist in global DNS) - proxy is not being used. HTTPToolkit says it Error: getaddrinfo ENOTFOUND a.b.c.d.existing.domain.zsz  and if it resolvable via local dns - it says it can't connect to it.
(it shouldn't connect to it directly - proxy must be used).

I tried changing proxy from http to socks (3proxy supports both) - HTTPToolkit still wants to either resolve a.b.c.d.existing.domain.zsz or connect to it directly even while it should it via configured proxy server.

How to solve this issue?

---

_Does this affect you too? Click the +:smiley: icon in the top right and add a :+1: on issues to help decide where HTTP Toolkit goes next, or [go take a look & vote on the most popular ideas so far](https://github.com/httptoolkit/feedback/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc)._

**Comment by pimterry** [comment]
Hmm, OK, interesting report!

I know for SOCKS specifically there's two different modes here, which currently HTTP Toolkit supports internally but doesn't expose, and probably should. Right now we default to socks5 (where the client resolves the IP, and then tells the proxy) but there it's also possible to configure socks5h instead (where the client sends the proxy the domain name, and the proxy resolves it itself).

It sounds like that would work for you, right? You could just select socks5h and use that for your proxy settings, and all resolution would be deferred to the proxy.

I'm a bit surprised that it doesn't work with plain HTTP though, since I would have expected that to work just fine. I've just tested this, and I can proxy traffic like this successfully (by setting up a dummy proxy server with Mockttp that sends a fixed response for proxied traffic to any possible address, using that as an HTTP proxy in HTTP Toolkit, and then making a request to `http://example.qweasd/`).

Do you have an easy way I can set up a repro for this so I can see the same behaviour you're seeing?

**Comment by intari** [comment]
hi,
plain http stared to work after full restart of both vm and host (and all network software running on it). 
I was able to repeat it - reset settings to basic, configure proxy in HTTPToolkit and in vm, try to use, it didn't worked until restart. but it work after it.

**Comment by pimterry** [comment]
I'm not sure I understand the full flow here, can you walk me through what you do, when it fails, what you change, and then exactly when it starts working, step by step?

One thing that I guess could affect this is caching - HTTP Toolkit does internally cache some DNS results. That's not supposed to cache failures, but it could be possible if something has gone wrong. Does it sound like that might cause this?

**Comment by intari** [comment]
yes, it's very possible it was something like DNS cache which wasn't cleared on proxy reconfiguration
